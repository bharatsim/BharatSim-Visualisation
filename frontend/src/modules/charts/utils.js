import { chartStyleConfig, getColor } from './chartStyleConfig';

function trasformDataForChart(csvData, xColumn, yColumns) {
  return {
    labels: csvData && csvData.data[xColumn],
    datasets:
      csvData &&
      yColumns.map((yColumn, index) => {
        return {
          ...chartStyleConfig,
          label: yColumn,
          borderColor: getColor(index),
          backgroundColor: getColor(index),
          data: csvData && csvData.data[yColumn],
        };
      }),
  };
}

function getYaxisNames(yColumns) {
  return yColumns.map((yColumn) => yColumn.name);
}

export { trasformDataForChart, getYaxisNames };
