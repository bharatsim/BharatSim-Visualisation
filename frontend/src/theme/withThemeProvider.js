import React from 'react';
import { MuiThemeProvider, CssBaseline } from '@material-ui/core';

import theme from './theme';

function withThemeProvider(WrappedComponent) {
  return function WithTheme(props) {
    return (
      <MuiThemeProvider theme={theme}>
        <CssBaseline />
        <WrappedComponent {...props} />
      </MuiThemeProvider>
    );
  };
}

export default withThemeProvider;
