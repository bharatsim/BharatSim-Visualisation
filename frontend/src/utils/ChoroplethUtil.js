import L from 'leaflet';
import chroma from 'chroma-js';

function highlightOnClick(e) {
  const layer = e.target;

  layer.setStyle({
    weight: 2,
    dashArray: '',
    opacity: 1,
    fillOpacity: 0.7,
  });

  if (!L.Browser.ie && !L.Browser.opera && !L.Browser.edge) {
    layer.bringToFront();
  }
}

function resetHighlight(geoJson) {
  return (e) => geoJson.current.resetStyle(e.target);
}

function zoomToFeature(map) {
  return (e) => map.fitBounds(e.target.getBounds());
}

function geoJSONStyle(idDataMap, mapLayerIdName, maxOfMeasure, scale) {
  const colors = Object.keys(scale)
    .sort((v1, v2) => Number(v1) - Number(v2))
    .map((key) => scale[key]);
  const getColor = chroma.scale(colors);
  return (feature) => {
    const gradientPercentage = idDataMap[feature.properties[mapLayerIdName]] / maxOfMeasure;
    return {
      color: '#1f2021',
      weight: 1,
      fillOpacity: 0.5,
      opacity: 0.7,
      fillColor: getColor(gradientPercentage),
    };
  };
}

export { highlightOnClick, resetHighlight, geoJSONStyle, zoomToFeature };
