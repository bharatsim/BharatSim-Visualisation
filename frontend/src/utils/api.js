import { contentTypes, httpMethods } from '../constants/fetch';
import { url as serviceURL } from './url';
import { fetchData, uploadData } from './fetch';

function headerBuilder({ contentType }) {
  return {
    'content-type': contentType,
  };
}

function formDataBuilder(data) {
  const formData = new FormData();
  data.forEach(({ name, value }) => formData.append(name, value));
  return formData;
}

const api = {
  saveDashboard: async ({ widgets, layout, dashboardId, name, count }) => {
    return uploadData({
      url: serviceURL.DASHBOARD_URL,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      data: JSON.stringify({
        dashboardData: { widgets, layout, dashboardId, name, count },
      }),
    });
  },

  addNewDashboard: async ({ name, projectId }) => {
    return uploadData({
      url: serviceURL.INSERT_DASHBOARD,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      data: JSON.stringify({
        dashboardData: { widgets: [], layout: [], name, count: 0, projectId },
      }),
      isCustomErrorHandler: true,
    });
  },

  getAllDashBoardByProjectId: async (projectId) =>
    fetchData({
      url: serviceURL.DASHBOARD_URL,
      query: { projectId, columns: ['name', '_id'] },
    }),

  getAllDashBoard: async () => fetchData({ url: serviceURL.DASHBOARD_URL }),

  uploadFileAndSchema: async ({ file, schema, dashboardId }) =>
    uploadData({
      url: serviceURL.DATA_SOURCES,
      data: formDataBuilder([
        { name: 'datafile', value: file },
        { name: 'schema', value: JSON.stringify(schema) },
        { name: 'dashboardId', value: dashboardId },
      ]),
      headers: headerBuilder({ contentType: contentTypes.FILE }),
    }),

  getCsvHeaders: async (dataSourceId) => fetchData({ url: serviceURL.getHeaderUrl(dataSourceId) }),

  getDatasources: async (dashboardId) =>
    fetchData({
      url: serviceURL.DATA_SOURCES,
      query: { dashboardId },
    }),

  getData: async (datasource, columns) =>
    fetchData({ url: serviceURL.getDataUrl(datasource), query: { columns } }),

  saveProject: async ({ id, ...data }) => {
    const requestObject = {
      url: serviceURL.PROJECT_URL,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      isCustomErrorHandler: true,
    };
    if (id) {
      return uploadData({
        ...requestObject,
        data: JSON.stringify({ projectData: { ...data, id } }),
        method: httpMethods.PUT,
      });
    }
    return uploadData({
      ...requestObject,
      data: JSON.stringify({ projectData: { ...data } }),
      method: httpMethods.POST,
    });
  },

  getProjects: async () => {
    return fetchData({
      url: serviceURL.PROJECT_URL,
    });
  },
  getProject: async (id) => {
    return fetchData({
      url: serviceURL.getProjectUrl(id),
    });
  },
};

export { api };
