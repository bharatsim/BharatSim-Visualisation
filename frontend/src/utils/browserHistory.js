// eslint-disable-next-line import/no-mutable-exports
let BrowserHistory;

function initHistory(history) {
  BrowserHistory = history;
}

function navigateToPreviousPage() {
  BrowserHistory.goBack();
}

function navigateToHome() {
  BrowserHistory.push('/');
}

function reloadPage() {
  // eslint-disable-next-line no-restricted-globals
  location.reload();
}

export { initHistory, navigateToPreviousPage, reloadPage, BrowserHistory, navigateToHome };
