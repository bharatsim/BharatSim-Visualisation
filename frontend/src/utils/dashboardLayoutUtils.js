export function getNewWidgetLayout(numberOfWidgetAdded, cols, count) {
  return createLayout({
    id: `widget-${count}`,
    xPosition: (numberOfWidgetAdded * 6) % cols,
    yPosition: Infinity, // puts it at the bottom
    width: 6,
    height: 2,
  });
}

function createLayout({ id, xPosition, yPosition, width, height }) {
  return {
    i: id,
    x: xPosition,
    y: yPosition,
    w: width,
    h: height,
  };
}
