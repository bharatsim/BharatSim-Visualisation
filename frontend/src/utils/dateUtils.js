const months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];

function getPeriod(hours) {
  return hours >= 12 ? 'PM' : 'AM';
}

function formatDate(dateString) {
  const date = new Date(dateString);
  if (date.toString() === 'Invalid Date') return '--';
  const DD = date.getDate();
  const MM = date.getMonth();
  const YYYY = date.getFullYear();
  const HH = date.getHours();
  const mm = date.getMinutes();

  return `${DD}-${months[MM]}-${YYYY} at ${HH % 12}:${mm < 10 ? `0${mm}` : mm} ${getPeriod(HH)}`;
}

function currentDate(){
  return new Date();
}

export { formatDate, currentDate };
