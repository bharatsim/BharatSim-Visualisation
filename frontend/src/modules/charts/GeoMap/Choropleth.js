import React, { useEffect, useMemo, useState } from 'react';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';
import { MapContainer, ScaleControl, TileLayer, ZoomControl } from 'react-leaflet';
import 'leaflet/dist/leaflet.css';
import Box from '@material-ui/core/Box';
import ResizeController from '../../../uiComponent/mapLayers/ResizeController';
import { api } from '../../../utils/api';
import useLoader from '../../../hook/useLoader';
import LoaderOrError from '../../loaderOrError/LoaderOrError';
import GeoJsonLayer from '../../../uiComponent/mapLayers/GeoJsonLayer';
import ColorScaleLegend from '../../../uiComponent/mapLayers/ColorScaleLegend';
import TimeSlider from '../../../uiComponent/TimeSlider';
import { transformChoroplethData } from '../../../utils/helper';
import { choroplethTypes } from '../../../constants/geoMap';
import DrillDownNavBreadcrumbs from '../../../uiComponent/DrillDownNavBreadcrumbs';
import useDeepCompareMemoize from '../../../hook/useDeepCompareMemoize';

const useStyles = makeStyles((theme) => ({
  fullWidthHeight: { height: '100%', width: '100%' },
  mapContainer: {
    height: `calc(100% - ${theme.spacing(11.5)}px )`,
  },
  breadcrumbContainer: {
    position: 'absolute',
    backgroundColor: theme.colors.grayScale['50'],
    zIndex: 100000,
    margin: theme.spacing(2),
    borderRadius: theme.spacing(1),
    padding: theme.spacing(0, 2),
  },
}));

const scale = { 0: '#FDF1D9', '0.5': '#FAA847', 1: '#EC3237' };

function Choropleth({ config }) {
  const classes = useStyles();
  const { dataSource, gisMeasure, choroplethConfig, sliderConfig } = config;
  const { choroplethType, mapLayerConfig } = choroplethConfig;
  const timeMetrics = sliderConfig?.timeMetrics;
  const {
    loadingState,
    message,
    startLoader,
    stopLoaderAfterError,
    stopLoaderAfterSuccess,
  } = useLoader();

  const [drillDownLevel, setDrillDownLevel] = useState(0);
  const [featureId, setFeatureId] = useState();
  const [levelFeatureMap, setLevelFeatureMap] = useState({ 0: undefined });
  const [data, setData] = useState();
  const [gisLayer, setGisLayer] = useState();
  const [timeSliderValue, setTimeSliderValue] = useState(1);

  const { mapLayer, mapLayerId, dataLayerId, referenceId } = mapLayerConfig[drillDownLevel];

  const mapLayerConfigDep = useDeepCompareMemoize(mapLayerConfig);
  useEffect(() => {
    setDrillDownLevel(0);
  }, [mapLayerConfigDep]);

  useEffect(() => {
    fetchAllData();
  }, [
    dataLayerId,
    gisMeasure,
    timeMetrics,
    dataLayerId,
    drillDownLevel,
    referenceId,
    mapLayer,
    choroplethConfig,
  ]);

  function onClickOfFeature(event) {
    const selectedFeature = event.target.feature.properties[mapLayerId];
    setFeatureId(selectedFeature);
    setDrillDownLevel((prevDrillDownLevel) => prevDrillDownLevel + 1);
    setLevelFeatureMap((prevMap) => ({ ...prevMap, [drillDownLevel + 1]: selectedFeature }));
  }

  const idMeasureMap = useMemo(
    () =>
      data
        ? transformChoroplethData(
            data[dataLayerId],
            data[gisMeasure],
            data[timeMetrics],
            timeSliderValue,
          )
        : {},
    [data, timeSliderValue],
  );

  const onErrorAction = {
    name: 'Retry',
    onClick: fetchAllData,
  };

  const isDrillDown = choroplethType === choroplethTypes.DRILL_DOWN;
  const shouldAddDrillDownCallback =
    choroplethType === choroplethTypes.DRILL_DOWN && drillDownLevel < mapLayerConfig.length - 1;

  const maxValue = data && data[gisMeasure] && Math.max(...data[gisMeasure]);
  const minValue = data && data[gisMeasure] && Math.min(...data[gisMeasure]);

  const breadcrumbsItems = Object.keys(levelFeatureMap)
    .sort()
    .slice(0, drillDownLevel + 1)
    .map((key, index) => {
      return {
        label: `Level ${index + 1}`,
        onClick: () => {
          const levelFeatureId = levelFeatureMap[key];
          setDrillDownLevel(index);
          setFeatureId(levelFeatureId);
        },
      };
    });

  return (
    <div className={timeMetrics ? classes.mapContainer : classes.fullWidthHeight}>
      <LoaderOrError loadingState={loadingState} message={message} errorAction={onErrorAction}>
        <div className={classes.fullWidthHeight} data-testid="map-container">
          <Box px={4}>
            {timeMetrics && data && data[timeMetrics] && (
              <Box>
                <TimeSlider
                  data={data[timeMetrics]}
                  setTimeSliderValue={setTimeSliderValue}
                  title={timeMetrics}
                  timeSliderValue={timeSliderValue}
                  sliderConfig={sliderConfig}
                />
              </Box>
            )}
          </Box>
          <MapContainer
            className={classes.fullWidthHeight}
            zoom={8}
            preferCanvas
            zoomControl={false}
          >
            {isDrillDown && (
              <Box className={classes.breadcrumbContainer}>
                <DrillDownNavBreadcrumbs items={breadcrumbsItems} />
              </Box>
            )}
            <TileLayer
              attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
              url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
            />
            <GeoJsonLayer
              mapLayer={gisLayer}
              idDataMap={idMeasureMap}
              mapLayerIdName={mapLayerId}
              measureName={gisMeasure}
              scale={scale}
              tick={timeSliderValue}
              onClickOfFeature={shouldAddDrillDownCallback ? onClickOfFeature : null}
              minOfMeasure={minValue}
              maxOfMeasure={maxValue}
            />
            <ResizeController />
            <ScaleControl />
            <ZoomControl position="bottomleft" />
            <ColorScaleLegend
              scale={scale}
              min={minValue}
              max={maxValue}
              title={gisMeasure}
              disablePercentageScale
            />
          </MapContainer>
        </div>
      </LoaderOrError>
    </div>
  );

  async function fetchAllData() {
    startLoader();
    try {
      await fetchGisLayer();
      await fetchAggregatedData();
      stopLoaderAfterSuccess();
    } catch (e) {
      stopLoaderAfterError('Unable to fetch data');
    }
  }

  async function fetchAggregatedData() {
    const { dataLayerId: parentDatalayerId } =
      drillDownLevel > 0 ? mapLayerConfig[drillDownLevel - 1] : {};
    const filter = drillDownLevel > 0 ? { propertyKey: parentDatalayerId, value: featureId } : null;
    const groupBy = [dataLayerId];
    const aggregations = { [gisMeasure]: 'sum' };

    if (timeMetrics) {
      groupBy.push(timeMetrics);
    }

    return api
      .getAggregatedData(dataSource, groupBy, aggregations, filter)
      .then(({ data: fetchedData }) => {
        setData(fetchedData);
      })
      .catch((error) => {
        throw error;
      });
  }

  async function fetchGisLayer() {
    const filter = drillDownLevel > 0 ? { propertyKey: referenceId, value: featureId } : null;
    return api
      .getAggregatedGeoJson(mapLayer, filter)
      .then(({ data: fetchedGisLayer }) => {
        setGisLayer(fetchedGisLayer);
      })
      .catch((error) => {
        throw error;
      });
  }
}

Choropleth.propTypes = {
  config: PropTypes.shape({
    dataSource: PropTypes.string.isRequired,
    gisMeasure: PropTypes.string.isRequired,
    choroplethConfig: PropTypes.shape({
      choroplethType: PropTypes.string,
      mapLayerConfig: PropTypes.arrayOf(
        PropTypes.shape({
          mapLayer: PropTypes.string,
          mapLayerId: PropTypes.string,
          dataLayerId: PropTypes.string,
          referenceId: PropTypes.string,
        }),
      ),
    }).isRequired,
    sliderConfig: PropTypes.shape({
      timeMetrics: PropTypes.string,
      strategy: PropTypes.string,
      stepSize: PropTypes.number,
    }).isRequired,
  }).isRequired,
};

export default React.memo(Choropleth);
