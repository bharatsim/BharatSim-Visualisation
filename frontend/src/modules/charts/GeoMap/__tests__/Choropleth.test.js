import React from 'react';
import { fireEvent } from '@testing-library/dom';
import { render } from '@testing-library/react';
import 'leaflet/dist/leaflet.css';
import { api } from '../../../../utils/api';
import withThemeProvider from '../../../../theme/withThemeProvider';
import Choropleth from '../Choropleth';

const mockGeoJson = {
  type: 'FeatureCollection',
  name: 'Gwalior',
  crs: {
    type: 'name',
    properties: { name: 'urn:ogc:def:crs:OGC:1.3:CRS84' },
  },
  features: [
    {
      type: 'Feature',
      properties: {
        AC_NO: 14,
      },
      geometry: {
        type: 'MultiPolygon',
        coordinates: [
          [
            [
              [78.264405697764801, 26.336678009382695],
              [78.274170968215685, 26.336251003181303],
              [78.28502017444157, 26.346887592588359],
            ],
          ],
        ],
      },
    },
  ],
};

jest.mock('../../../../utils/api', () => ({
  api: {
    getAggregatedGeoJson: jest
      .fn()
      .mockImplementation(() => Promise.resolve({ data: mockGeoJson })),
    getAggregatedData: jest.fn().mockImplementation(() =>
      Promise.resolve({
        data: { regionId: [1, 2, 3], infected: [2, 4, 6], timeTick: [1, 2, 3] },
      }),
    ),
  },
}));

jest.mock('react-leaflet', () => ({
  MapContainer: ({ children }) => (
    <>
      MapContainer
      {children}
    </>
  ),
  TileLayer: () => <>TileLayer</>,
  ScaleControl: () => <>ScaleControl</>,
  ZoomControl: () => <>ZoomControl</>,
}));

jest.mock('../../../../uiComponent/mapLayers/HeatMapLayer', () => (props) => (
  <>
    HeatMapLayer
    {JSON.stringify(props)}
  </>
));

jest.mock('../../../../uiComponent/mapLayers/ViewController', () => (props) => (
  <>
    ViewAndZoomLayer
    {JSON.stringify(props)}
  </>
));

jest.mock('../../../../uiComponent/mapLayers/ColorScaleLegend', () => (props) => (
  <>
    ColorScaleLegend
    {JSON.stringify(props)}
  </>
));

jest.mock('../../../../uiComponent/mapLayers/ResizeController', () => (props) => (
  <>
    Resize controller
    {JSON.stringify(props)}
  </>
));

jest.mock('../../../../uiComponent/mapLayers/GeoJsonLayer', () => (props) => (
  <>
    GeoJsonLayer
    <button
      onClick={() => {
        props.onClickOfFeature({ target: { feature: { properties: { mapLayerId1: 'region1' } } } });
      }}
    >
      feature1
    </button>
    {JSON.stringify(props)}
  </>
));

const config = {
  dataSource: 'test.csv',
  gisMeasure: 'infected',
  choroplethConfig: {
    choroplethType: 'singleLevel',
    mapLayerConfig: [
      {
        mapLayer: 'mapLayer.geoJson',
        mapLayerId: 'mapLayerId1',
        dataLayerId: 'regionId',
        referenceId: '',
      },
    ],
  },
  sliderConfig: {
    timeMetrics: 'timeTick',
    strategy: 'defaultIntervals',
    stepSize: 1,
  },
};

describe('Choropleth', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });
  beforeEach(() => {
    jest.runAllTimers();
  });
  afterEach(() => {
    jest.clearAllMocks();
    jest.clearAllMocks();
  });

  const ChoroplethWithProvider = withThemeProvider(Choropleth);

  it('should create a choropleth component', async () => {
    const { container, findByTestId } = render(
      <ChoroplethWithProvider
        config={{
          dataSource: 'test.csv',
          gisMeasure: 'infected',
          choroplethConfig: config.choroplethConfig,
          sliderConfig: config.sliderConfig,
        }}
      />,
    );

    await findByTestId('map-container');

    expect(container).toMatchSnapshot();
  });

  it('should create a choropleth component without time slider', async () => {
    const { container, findByTestId } = render(
      <ChoroplethWithProvider
        config={{
          dataSource: 'test.csv',
          gisMeasure: 'infected',
          choroplethConfig: config.choroplethConfig,
          sliderConfig: {},
        }}
      />,
    );

    await findByTestId('map-container');

    expect(container).toMatchSnapshot();
  });

  it('should call fetch data on render with time Metrics if time metric is present with aggregated data', async () => {
    api.getAggregatedData.mockImplementation(() =>
      Promise.resolve({
        data: { regionId: [1, 2, 3], infected: [2, 4, 6], day: [1, 2, 3] },
      }),
    );

    const { findByTestId } = render(
      <ChoroplethWithProvider
        config={{
          dataSource: 'test.csv',
          gisMeasure: 'infected',
          choroplethConfig: { ...config.choroplethConfig, choroplethType: 'drillDown' },
          sliderConfig: config.sliderConfig,
        }}
      />,
    );

    await findByTestId('map-container');

    expect(api.getAggregatedData).toHaveBeenCalledWith(
      'test.csv',
      ['regionId', 'timeTick'],
      {
        infected: 'sum',
      },
      null,
    );
  });

  it('should show breadcrumbs when drill down map is selected', async () => {
    const { findByTestId, getByText } = render(
      <ChoroplethWithProvider
        config={{
          dataSource: 'test.csv',
          gisMeasure: 'infected',
          choroplethConfig: { ...config.choroplethConfig, choroplethType: 'drillDown' },
          sliderConfig: config.sliderConfig,
        }}
      />,
    );

    await findByTestId('map-container');

    expect(getByText('Level 1')).toBeInTheDocument();
  });

  it('should show breadcrumbs for Top level and  level 2', async () => {
    api.getAggregatedGeoJson.mockImplementation(() => Promise.resolve({ data: mockGeoJson }));
    api.getAggregatedData.mockImplementation(() =>
      Promise.resolve({
        data: { regionId: [1, 2, 3], infected: [2, 4, 6], regionId2: [1, 2, 3] },
      }),
    );
    const choroplethConfig = {
      choroplethType: 'drillDown',
      mapLayerConfig: [
        {
          mapLayer: 'mapLayer.geoJson',
          mapLayerId: 'mapLayerId1',
          dataLayerId: 'regionId',
          referenceId: '',
        },
        {
          mapLayer: 'mapLayer2.geoJson',
          mapLayerId: 'mapLayerId2',
          dataLayerId: 'regionId2',
          referenceId: 'mapLayerId1',
        },
      ],
    };

    const { getByText, findByTestId } = render(
      <ChoroplethWithProvider
        config={{
          dataSource: 'test.csv',
          gisMeasure: 'infected',
          choroplethConfig,
          sliderConfig: {},
        }}
      />,
    );

    await findByTestId('map-container');

    fireEvent.click(getByText('feature1').closest('button'));

    await findByTestId('map-container');

    expect(getByText('Level 1')).toBeInTheDocument();
    expect(getByText('Level 2')).toBeInTheDocument();
  });

  it('should fetch top level data on click of Top level breadcrumb label', async () => {
    api.getAggregatedGeoJson.mockImplementation(() => Promise.resolve({ data: mockGeoJson }));
    api.getAggregatedData.mockImplementation(() =>
      Promise.resolve({
        data: { regionId: [1, 2, 3], infected: [2, 4, 6], regionId2: [1, 2, 3] },
      }),
    );
    const choroplethConfig = {
      choroplethType: 'drillDown',
      mapLayerConfig: [
        {
          mapLayer: 'mapLayer.geoJson',
          mapLayerId: 'mapLayerId1',
          dataLayerId: 'regionId',
          referenceId: '',
        },
        {
          mapLayer: 'mapLayer2.geoJson',
          mapLayerId: 'mapLayerId2',
          dataLayerId: 'regionId2',
          referenceId: 'mapLayerId1',
        },
      ],
    };

    const { getByText, findByTestId } = render(
      <ChoroplethWithProvider
        config={{
          dataSource: 'test.csv',
          gisMeasure: 'infected',
          choroplethConfig,
          sliderConfig: {},
        }}
      />,
    );

    await findByTestId('map-container');

    fireEvent.click(getByText('feature1').closest('button'));

    await findByTestId('map-container');
    const topLevelButton = getByText('Level 1').closest('button');
    fireEvent.click(topLevelButton);

    await findByTestId('map-container');

    expect(api.getAggregatedData).toHaveBeenCalledWith(
      'test.csv',
      ['regionId'],
      {
        infected: 'sum',
      },
      null,
    );
    expect(api.getAggregatedGeoJson).toHaveBeenCalledWith('mapLayer.geoJson', null);
  });

  it('should rerender when prop value changes', async () => {
    api.getAggregatedData.mockImplementation(() =>
      Promise.resolve({
        data: {
          regionId: [1, 2, 3],
          infected: [2, 4, 6],
          day: [1, 2, 3],
          infectedNew: [1, 2, 3],
        },
      }),
    );

    const { findByTestId, rerender } = render(
      <ChoroplethWithProvider
        config={{
          dataSource: 'test.csv',
          gisMeasure: 'infected',
          choroplethConfig: config.choroplethConfig,
          sliderConfig: config.sliderConfig,
        }}
      />,
    );

    await findByTestId('map-container');
    expect(api.getAggregatedData).toHaveBeenCalledWith(
      'test.csv',
      ['regionId', 'timeTick'],
      { infected: 'sum' },
      null,
    );

    rerender(
      <ChoroplethWithProvider
        config={{
          dataSource: 'test.csv',
          gisMeasure: 'infectedNew',
          choroplethConfig: config.choroplethConfig,
          sliderConfig: config.sliderConfig,
        }}
      />,
    );

    await findByTestId('map-container');

    expect(api.getAggregatedData).toHaveBeenLastCalledWith(
      'test.csv',
      ['regionId', 'timeTick'],
      { infectedNew: 'sum' },
      null,
    );
  });

  it('should call fetch data on click of retry button', async () => {
    api.getAggregatedData.mockRejectedValueOnce('error');
    const { findByText, getByText, findByTestId } = render(
      <ChoroplethWithProvider
        config={{
          dataSource: 'test.csv',
          gisMeasure: 'infected',
          choroplethConfig: config.choroplethConfig,
          sliderConfig: config.sliderConfig,
        }}
      />,
    );

    await findByText('Unable to fetch data');

    const retryButton = getByText('Retry');

    fireEvent.click(retryButton);

    await findByTestId('map-container');

    expect(api.getAggregatedGeoJson).toHaveBeenCalledWith('mapLayer.geoJson', null);
    expect(api.getAggregatedData).toHaveBeenCalledWith(
      'test.csv',
      ['regionId', 'timeTick'],
      { infected: 'sum' },
      null,
    );
  });

  it('should show error if get data api fail to load for map layer', async () => {
    api.getAggregatedGeoJson.mockImplementation(() => Promise.reject('error'));
    const { findByText, getByText } = render(
      <ChoroplethWithProvider
        config={{
          dataSource: 'test.csv',
          gisMeasure: 'infected',
          choroplethConfig: config.choroplethConfig,
          sliderConfig: config.sliderConfig,
        }}
      />,
    );

    await findByText('Unable to fetch data');

    expect(getByText('Unable to fetch data')).toBeInTheDocument();
  });

  it('should show error if get data api fail to aggregated data', async () => {
    api.getAggregatedData.mockImplementation(() => Promise.reject('error'));
    api.getAggregatedGeoJson.mockImplementation(() => Promise.resolve({ data: mockGeoJson }));
    const { findByText, getByText } = render(
      <ChoroplethWithProvider
        config={{
          dataSource: 'test.csv',
          gisMeasure: 'infected',
          choroplethConfig: { ...config.choroplethConfig, choroplethType: 'drillDown' },
          sliderConfig: config.sliderConfig,
        }}
      />,
    );

    await findByText('Unable to fetch data');

    expect(getByText('Unable to fetch data')).toBeInTheDocument();
  });

  it('should call apis to fetch level2 data and mapLayer on click of feature', async () => {
    api.getAggregatedGeoJson.mockImplementation(() => Promise.resolve({ data: mockGeoJson }));
    api.getAggregatedData.mockImplementation(() =>
      Promise.resolve({
        data: { regionId: [1, 2, 3], infected: [2, 4, 6], regionId2: [1, 2, 3] },
      }),
    );

    const choroplethConfig = {
      choroplethType: 'drillDown',
      mapLayerConfig: [
        {
          mapLayer: 'mapLayer.geoJson',
          mapLayerId: 'mapLayerId1',
          dataLayerId: 'regionId',
          referenceId: '',
        },
        {
          mapLayer: 'mapLayer2.geoJson',
          mapLayerId: 'mapLayerId2',
          dataLayerId: 'regionId2',
          referenceId: 'mapLayerId1',
        },
      ],
    };

    const { getByText, findByTestId } = render(
      <ChoroplethWithProvider
        config={{
          dataSource: 'test.csv',
          gisMeasure: 'infected',
          choroplethConfig,
          sliderConfig: {},
        }}
      />,
    );

    await findByTestId('map-container');

    expect(api.getAggregatedData).toHaveBeenCalledWith(
      'test.csv',
      ['regionId'],
      {
        infected: 'sum',
      },
      null,
    );
    expect(api.getAggregatedGeoJson).toHaveBeenCalledWith('mapLayer.geoJson', null);

    fireEvent.click(getByText('feature1').closest('button'));

    await findByTestId('map-container');

    expect(api.getAggregatedData).toHaveBeenLastCalledWith(
      'test.csv',
      ['regionId2'],
      {
        infected: 'sum',
      },
      { propertyKey: 'regionId', value: 'region1' },
    );
    expect(api.getAggregatedGeoJson).toHaveBeenLastCalledWith('mapLayer2.geoJson', {
      propertyKey: 'mapLayerId1',
      value: 'region1',
    });
  });
});
