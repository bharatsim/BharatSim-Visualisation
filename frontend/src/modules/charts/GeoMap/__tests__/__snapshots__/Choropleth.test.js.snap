// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Choropleth should create a choropleth component 1`] = `
<div>
  <div
    class="makeStyles-mapContainer-2"
  >
    <div
      class="makeStyles-fullWidthHeight-1"
      data-testid="map-container"
    >
      <div
        class="MuiBox-root MuiBox-root-6"
      >
        <div
          class="MuiBox-root MuiBox-root-7"
        >
          <div
            class="makeStyles-root-8"
          >
            <div
              class="MuiBox-root MuiBox-root-10"
            >
              <p
                class="MuiTypography-root makeStyles-title-9 MuiTypography-body1"
              >
                timeTick: 1
              </p>
            </div>
            <span
              class="MuiSlider-root MuiSlider-colorPrimary MuiSlider-marked"
            >
              <span
                class="MuiSlider-rail"
              />
              <span
                class="MuiSlider-track"
                style="left: 0%; width: 0%;"
              />
              <input
                type="hidden"
                value="1"
              />
              <span
                class="MuiSlider-mark MuiSlider-markActive"
                data-index="0"
                style="left: 0%;"
              />
              <span
                aria-hidden="true"
                class="MuiSlider-markLabel MuiSlider-markLabelActive"
                data-index="0"
                style="left: 0%;"
              >
                1
              </span>
              <span
                class="MuiSlider-mark"
                data-index="1"
                style="left: 50%;"
              />
              <span
                class="MuiSlider-mark"
                data-index="2"
                style="left: 100%;"
              />
              <span
                aria-hidden="true"
                class="MuiSlider-markLabel"
                data-index="2"
                style="left: 100%;"
              >
                3
              </span>
              <span
                aria-labelledby="Time Slider"
                aria-orientation="horizontal"
                aria-valuemax="3"
                aria-valuemin="1"
                aria-valuenow="1"
                class="MuiSlider-thumb MuiSlider-thumbColorPrimary"
                data-index="0"
                role="slider"
                style="left: 0%;"
                tabindex="0"
              />
            </span>
          </div>
        </div>
      </div>
      MapContainer
      TileLayer
      GeoJsonLayer
      <button>
        feature1
      </button>
      {"mapLayer":{"type":"FeatureCollection","name":"Gwalior","crs":{"type":"name","properties":{"name":"urn:ogc:def:crs:OGC:1.3:CRS84"}},"features":[{"type":"Feature","properties":{"AC_NO":14},"geometry":{"type":"MultiPolygon","coordinates":[[[[78.2644056977648,26.336678009382695],[78.27417096821569,26.336251003181303],[78.28502017444157,26.34688759258836]]]]}}]},"idDataMap":{"1":2},"mapLayerIdName":"mapLayerId1","measureName":"infected","scale":{"0":"#FDF1D9","1":"#EC3237","0.5":"#FAA847"},"tick":1,"onClickOfFeature":null,"minOfMeasure":2,"maxOfMeasure":6}
      Resize controller
      {}
      ScaleControl
      ZoomControl
      ColorScaleLegend
      {"scale":{"0":"#FDF1D9","1":"#EC3237","0.5":"#FAA847"},"min":2,"max":6,"title":"infected","disablePercentageScale":true}
    </div>
  </div>
</div>
`;

exports[`Choropleth should create a choropleth component without time slider 1`] = `
<div>
  <div
    class="makeStyles-fullWidthHeight-16"
  >
    <div
      class="makeStyles-fullWidthHeight-16"
      data-testid="map-container"
    >
      <div
        class="MuiBox-root MuiBox-root-21"
      />
      MapContainer
      TileLayer
      GeoJsonLayer
      <button>
        feature1
      </button>
      {"mapLayer":{"type":"FeatureCollection","name":"Gwalior","crs":{"type":"name","properties":{"name":"urn:ogc:def:crs:OGC:1.3:CRS84"}},"features":[{"type":"Feature","properties":{"AC_NO":14},"geometry":{"type":"MultiPolygon","coordinates":[[[[78.2644056977648,26.336678009382695],[78.27417096821569,26.336251003181303],[78.28502017444157,26.34688759258836]]]]}}]},"idDataMap":{"1":2,"2":4,"3":6},"mapLayerIdName":"mapLayerId1","measureName":"infected","scale":{"0":"#FDF1D9","1":"#EC3237","0.5":"#FAA847"},"tick":1,"onClickOfFeature":null,"minOfMeasure":2,"maxOfMeasure":6}
      Resize controller
      {}
      ScaleControl
      ZoomControl
      ColorScaleLegend
      {"scale":{"0":"#FDF1D9","1":"#EC3237","0.5":"#FAA847"},"min":2,"max":6,"title":"infected","disablePercentageScale":true}
    </div>
  </div>
</div>
`;
